cpp/dolfin/io/XDMFFile.cpp

```diff
@@ -529,14 +573,53 @@ void XDMFFile::write(const function::Function& u, double time_step)
   }
 
   // Get function::Function data values and shape
-  std::vector<double> data_values;
+  std::vector<PetscScalar> data_values;
   bool cell_centred = has_cell_centred_data(u);
 
   if (cell_centred)
     data_values = get_cell_data_values(u);
   else
     data_values = get_point_data_values(u);
 
+  // Add attribute DataItem node and write data
+  std::int64_t width = get_padded_width(u);
+  assert(data_values.size() % width == 0);
+  std::int64_t num_values
+      = cell_centred ? mesh.num_entities_global(mesh.topology().dim())
+                     : mesh.num_entities_global(0);
+
+#ifdef PETSC_USE_COMPLEX
+  std::vector<std::string> components = {"real", "imag"};
+  for (const std::string component : components)
+  {
+    std::string attr_name = component + "_" + u.name();
+    // Add attribute node
+    pugi::xml_node attribute_node = mesh_node.append_child("Attribute");
+    assert(attribute_node);
+    attribute_node.append_attribute("Name") = attr_name.c_str();
+    attribute_node.append_attribute("AttributeType")
+        = rank_to_string(u.value_rank()).c_str();
+    attribute_node.append_attribute("Center") = cell_centred ? "Cell" : "Node";
+
+    const std::string dataset_name
+        = "/VisualisationVector/" + component + "/" + std::to_string(_counter);
+
+    // FIXME: Avoid copies by writing directly a compound data
+    std::vector<double> component_data_values(data_values.size());
+    for (unsigned int i = 0; i < data_values.size(); i++)
+    {
+      if (component == components[0])
+        component_data_values[i] = data_values[i].real();
+      else if (component == components[1])
+        component_data_values[i] = data_values[i].imag();
+    }
+
+    // Add data item of component
+    add_data_item(_mpi_comm.comm(), attribute_node, h5_id, dataset_name,
+                  component_data_values, {num_values, width});
+  }
+
+#else
```

Comment:
The same consideration as above - is sharing more code with the non-complex case possible? It is probably beneficial to leave this out for the `H5Sselect_hyperslab` issue though.

Reply:
This whole `#ifdef PETSC_USE_COMPLEX ...  #endif` can go away with the correct use of H5Sselect_hyperslab. We have to specify the `stride=2`, and `offset = 0 or 1` depending on the component (real and imaginary). 
Just have figure out how this information should reach `HDF5Interface::write_dataset`.